org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/RecordPattern.java

```diff
@@ -113,85 +115,92 @@ public boolean coversType(TypeBinding t) {
 		return true;
 	}
 	@Override
-	public void resolveWithExpression(BlockScope scope, Expression exp) {
-		this.expression = exp;
-		if (shouldInitiateRecordTypeInference()) {
-			LocalVariableBinding localVariableBinding = exp.localVariableBinding();
-			TypeBinding type1 = localVariableBinding != null && localVariableBinding.type != null ?
-					localVariableBinding.type : exp.resolvedType;
-			if (type1 instanceof ReferenceBinding) {
-				ReferenceBinding binding = inferRecordParameterization(scope, (ReferenceBinding) type1);
-				if (binding == null || !binding.isValidBinding()) {
-					scope.problemReporter().cannotInferRecordPatternTypes(this);
-				    this.resolvedType = null;
-				    return;
-				}
-				this.resolvedType = binding;
-				setAccessorsPlusInfuseInferredType(scope);
-			} else {
-				// TODO: which scenarios? eg? if found add the code here from resolveType();
-			}
-		}
-	}
-	@Override
 	public TypeBinding resolveAtType(BlockScope scope, TypeBinding u) {
 		for (Pattern p : this.patterns) {
 			p.resolveAtType(scope, u);
 		}
 		return this.resolvedType;
 	}
 	@Override
-	public TypeBinding resolveTypeWithBindings(LocalVariableBinding [] bindings, BlockScope scope) {
+	public TypeBinding resolveTypeWithBindings(LocalVariableBinding[] bindings, BlockScope scope) {
 		scope.include(bindings);
 		try {
-			if (this.resolvedType != null)
-				return this.resolvedType;
+			return resolveType(scope);
+		} finally {
+			scope.exclude(bindings);
+		}
+	}
 
-			this.type.bits |= ASTNode.IgnoreRawTypeCheck;
-			this.resolvedType = this.type.resolveType(scope);
+	@Override
+	public void setExpectedType(TypeBinding expectedType) {
+		this.expectedType = expectedType;
+	}
 
-			if (this.resolvedType == null || !this.resolvedType.isValidBinding()) {
-				return this.resolvedType;
-			}
+	@Override
+	public TypeBinding expectedType() {
+		return this.expectedType;
+	}
 
-			getSecretVariable(scope, this.resolvedType);
+	@Override
+	public TypeBinding resolveType(BlockScope scope) {
+
+		if (this.resolvedType != null)
+			return this.resolvedType;
 
-			// check whether the give type reference is a record
-			// check whether a raw type is being used in pattern types
-			// check whether the pattern signature matches that of the record declaration
-			if (!this.resolvedType.isRecord()) {
-				scope.problemReporter().unexpectedTypeinRecordPattern(this.resolvedType, this.type);
-				return this.resolvedType;
+		this.type.bits |= ASTNode.IgnoreRawTypeCheck;
+		this.resolvedType = this.type.resolveType(scope);
+
+		if (!this.resolvedType.isRecord()) {
```

Comment:
@srikanth-sankaran  prior to this change, this block sat in `resolveTypeWithBindings` where the case of `this.resolvedType == null` had an early exit prior to it. It's a long shot but could this change be the cause of the NPE reported in https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2399#issuecomment-2315632390 where _some call_ `this.resolvedType.isRecord()` _somewhere_ blew up.